File: Code/Code2.c
/*
Scrivere una funzione che inverta una coda, ovvero produca una coda invertendo l’ordine
degli elementi della coda di partenza. Suggerimento: utilizzare una pila.
*/
#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>
#include <string.h>

typedef struct nodo{
    int valore;
    struct nodo* next;
}Nodo;

void enqueue(Nodo** head,Nodo** tail,Nodo* element){
    if(isEmpty(*head)) *head = element;
    else    (*tail)->next = element;
    *tail = element;
    element->next = NULL;
}

Nodo* dequeue(Nodo** head,Nodo** tail){
    Nodo* ret = *head;

    if(isEmpty(*head)) return NULL;
    else    *head = ret->next;

    if(*head == NULL) *tail = NULL;
    return ret;
}

int isEmpty(Nodo *struttura){
    if(struttura == NULL) return 1;
    return 0;
}

void stampaCoda(struct Nodo** head, struct Nodo** tail){
    while(!isEmpty(*head)){
        printf("%d ", dequeue(head,tail)->valore);
    }
}

void push(Nodo** head, Nodo* element){
    if(isEmpty(*head)){
        *head=element;
        element->next=NULL;
    }else{
        element->next=*head;
        *head= element;
    }
}

Nodo* pop(Nodo **head){
    Nodo* ret= *head;
    if(isEmpty(*head)){
        return NULL;
    }else{
        *head= ret->next;
    }
    return ret;
}

void invertiCoda(Nodo** head,Nodo** tail){
    Nodo* stack=NULL;
    while(!isEmpty(*head)){
        push(&stack,dequeue(head,tail));
    }
    while(!isEmpty(stack)){
        enqueue(head,tail,pop(&stack));
    }
}

int main(){
    Nodo* headCoda;
    Nodo* tailCoda;
    Nodo* element;
    char scelta;

    headCoda=NULL;
    tailCoda=NULL;
    
    do{
        element=(Nodo*)malloc(sizeof(Nodo));
        printf("Inserire un numero: ");
        fflush(stdin);
        scanf("%d",&element->valore);
        enqueue(&headCoda,&tailCoda,element);
        printf("Inserire n se si vuole smettere di inserire: ");
        fflush(stdin);
        scanf("%c",&scelta);
    }while(scelta!='n');

    invertiCoda(&headCoda,&tailCoda);
    stampaCoda(&headCoda,&tailCoda);
    return 0;
}

File: Code/Code2.py
'''
Scrivere una funzione che inverta una coda, ovvero produca una coda invertendo l’ordine
degli elementi della coda di partenza. Suggerimento: utilizzare una pila.
'''
def enqueue(queue, element):                    #Funzione per caricare la coda
    queue.append(element)


def dequeue(queue):                             #Funzione per svuotare la coda
    return queue.pop(0)


def push(stack, element):                       #Funzione per il caricamento delle pile 
    stack.append(element)


def pop(stack):                                 #Funzione per togliere elementi dalle pile 
    return stack.pop()

def main():
    coda=[]                                     #coda che viene caricata dall'utente
    pilaTemp=[]                                 #pila d'appoggio per l'inversione 

    num=int(input("Inserire un numero: "))      #Inseriemnto degli input 
    enqueue(coda,num)
    while input("Vuoi continuare (Inserire n per terminare l'input): ")!='n':
        num=int(input("Inserire un numero: "))
        enqueue(coda,num)
    
    for _ in range(len(coda)):                  #La coda viene svuotata e la pila viene riempita
        push(pilaTemp,dequeue(coda))
    
    for _ in range(len(pilaTemp)):              #La pila viene svuota e la coda viene riempita, ma con l'ordine inverso 
        enqueue(coda,pop(pilaTemp))
    
    for element in coda:                        #Stampa di tutti gli elementi inseriti nella coda
        print(element)


if __name__=="__main__":
    main()

    
